Add Autocomplete to chat text box and track visible code

hassoncs opened the pull request:
Add support for Fill-in-the-Middle (FIM) completion in the chat text area, providing ghost text suggestions based on the currently visible code context. Includes new message types for requesting visible code and chat completions, updated webview handlers, enhanced UI components with Tab acceptance and Escape cancellation, and new service modules for context tracking and autocomplete logic.

![2025-11-24 10 19 38](https://github.com/user-attachments/assets/d948f766-d4ca-4919-8de1-21c4e3c0100b)
<img width="455" height="398" alt="CleanShot 2025-12-08 at 20 04 28" src="https://github.com/user-attachments/assets/ba49f0a3-058a-4259-a921-02c535765609" />

hassoncs committed:
Add chat autocomplete functionality and related types

- Implemented chat text area autocomplete feature with new message handling for chat completion requests.
- Introduced new types for chat completion requests and results in `types.ts`.
- Updated `ExtensionMessage` and `WebviewMessage` to support chat completion messages.
- Styled ghost text for autocomplete suggestions in CSS.
- Refactored `ChatTextArea` component to utilize a custom hook for handling ghost text input and display.
- Enhanced HTML escaping utility to be exported for broader use.

hassoncs committed:
Add chat autocomplete settings and integration

- Added `enableChatAutocomplete` option to ghost service settings schema.
- Updated `ChatTextArea` to utilize the new autocomplete feature based on the settings.
- Enhanced `useChatGhostText` hook to conditionally request completions based on the new setting.
- Implemented UI components for toggling chat autocomplete in the settings view.
- Updated localization files to include translations for the new chat autocomplete feature.

hassoncs committed:
feat(chat): add word-by-word ghost text acceptance in chat

- Enhance Tab key to accept full ghost text completion
- Add ArrowRight key to accept next word only, mimicking VS Code behavior
- Refactor key handling logic for better cursor and selection checks
- Remove unused inputValue/setInputValue dependencies from hook
- Add helper functions for word extraction and text insertion

Add test directory for useChatGhostText hook validation

hassoncs committed:
feat(ghost): add filtering for sensitive files in visible code tracking

Integrate RooIgnoreController into VisibleCodeTracker to respect .kilocodeignore patterns and security concerns, preventing sensitive files from being included in chat autocompletion context. Add corresponding tests for filtering behavior.

hassoncs committed:
changeset

markijbema reviewed (dismissed):
I really like this idea, but it adds a lot of changes which shouldn't be necessary; we can use a completion provider in the chatbox, so there isnt a change necessary in the chattextarea, or any roo-merge adjecent areas.

Also I don't think we need the visible code tracker, but we can use the already existing  context gathering code

hassoncs committed:
refactor(test): consolidate chat autocomplete tests and update visible code tracker test description

- Merge multiple test cases into fewer blocks for better readability
- Update test description to reflect broader security filtering scope
- Remove redundant helper functions in favor of direct assertions

hassoncs committed:
fix changeset

hassoncs committed:
fix(chat): improve FIM availability check to validate credentials

- Update isFimAvailable() to check hasValidCredentials() instead of loaded
- Add test case for isFimAvailable when model is not loaded

hassoncs commented on webview-ui/src/components/chat/hooks/useChatGhostText.ts:
```diff
@@ -0,0 +1,174 @@
+	// Match: optional leading whitespace + non-whitespace characters + optional trailing whitespace
+	// This captures " word " or "word " or " word" as complete units
+	const match = text.match(/^(\s*\S+\s*)/)
+	if (match) {
```
Yes, it's deprecated, but because this is VS Code, we know for sure it's going to work. So let's leave this alone since it works the same on the roo side.

markijbema commented on src/core/webview/webviewMessageHandler.ts:
```diff
@@ -3692,7 +3692,19 @@ export const webviewMessageHandler = async (
 			}
 			break
+		} // kilocode_change start: Chat text area FIM autocomplete
+		case "requestChatCompletion": {
```
Yeah you just need to register it. If you register * it does this, as well as autocomplete in the commit messages. Clearly we don't want that, but I suppose you can then also do it more granular

hassoncs committed:
Enable chat autocomplete by default

hassoncs committed:
Address code review suggestions

- Rename test files to use .spec.ts convention
- Remove debug console.log statements from ChatTextAreaAutocomplete
